File: Java/Problem4.java

public String shortestPalindrome(String s) {
        String reverse = new StringBuilder(s).reverse().toString();
        String combinedString = s + "." + reverse;

        int[] lps = new int[combinedString.length()];

        for (int i = 1; i < combinedString.length(); i++) {
            int pattern = lps[i - 1];
            while (pattern > 0 && combinedString.charAt(i) != combinedString.charAt(pattern)) {
                pattern = lps[pattern - 1];
            }
            if (combinedString.charAt(i) == combinedString.charAt(pattern)) {
                pattern++;
            }
            lps[i] = pattern;
        }

        String charsToAdd = s.substring(lps[combinedString.length() - 1]);
        String prefix = new StringBuilder(charsToAdd).reverse().toString();

        return prefix + s;
    }

File: Python/Problem4.py

def shortestPalindrome(self, s: str) -> str:
    reverse = s[::-1]
    combined_string = s + '.' + reverse

    lps = [0] * len(combined_string)

    for i in range(1, len(combined_string)):
        pattern = lps[i - 1]
        while pattern > 0 and combined_string[i] != combined_string[pattern]:
            pattern = lps[pattern - 1]
        if combined_string[i] == combined_string[pattern]:
            pattern += 1
        lps[i] = pattern

    chars_to_add = s[lps[-1]:]
    return chars_to_add[::-1] + s
